### PR TITLE
WIP - Added support for using an existing ApiGateway

### LINKED ISSUE #3078: Deploy many micro/nano services to one API Gateway
# This is a Feature Proposal

## Description

I'd like to deploy many microservices to one API Gateway, but currently I'm not able to as the template is creating a new Gateway for every service. Would it be possible to change the behaviour? I may create a Pull Request for this if you agree

Topic on the forum: http://forum.serverless.com/t/multiple-services-behind-a-single-api-gateway/191

### PR DESCRIPTION
**Pull request by eahefnawy**
## What did you implement:

Closes #3078
Added support for specifying an existing ApiGateway, and hence sharing this ApiGateway with multiple services. [Inspired by @HyperBrain ideas](https://github.com/serverless/serverless/issues/3078#issuecomment-310690473) 😊 

## How did you implement it:
Users need to provide a Rest Api Id and Rest Api Root Resource Id for that existing ApiGateway. The framework will use this configuration instead of creating a rest api from scratch.

## How can we verify it:
You can provide the new configuration by providing strings directly:

```yml
service:
  name: sls-shared-api

provider:
  name: aws
  runtime: nodejs6.10
  apiGatewayRestApiId: 6fyzt1pfpk # can be found in aws console
  apiGatewayRestApiRootResourceId: z5d4qh4oqi # can be found in aws console

functions:
  hello:
    handler: handler.hello
    events:
      - http: POST hello
      - http: POST hello/again
```

or by using an intrinsic function such as Ref, GetAttr, ImportValue...etc:

```yml
service:
  name: sls-shared-api

provider:
  name: aws
  runtime: nodejs6.10
  apiGatewayRestApiId:
    Ref: MySharedApiGatewayRestApi
  apiGatewayRestApiRootResourceId: { 'Fn::GetAtt': [MySharedApiGatewayRestApi, 'RootResourceId'] }
    

functions:
  hello:
    handler: handler.hello
    events:
      - http: POST hello
      - http: POST hello/again

resources:
  Resources:
    MySharedApiGatewayRestApi:
      Type: AWS::ApiGateway::RestApi
      Properties:
        Name: dev-my-shared-api-gateway
```

## Todos:

- [ ] Write tests
- [ ] Write documentation
- [ ] Fix linting errors
- [ ] Make sure code coverage hasn't dropped
- [ ] Provide verification config / commands / resources
- [ ] Enable "Allow edits from maintainers" for this PR
- [ ] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by eahefnawy** [comment]
One issue I'm having with this PR is that the user has to specify a Rest API Root Resource Id. I was hopping specifying a Rest API Id would be enough, but unfortunately the framework would need to figure out the root resource of that existing API somehow. One solution is to fetch and filter all the APIG resources using the AWS SDK by providing the Rest API Id, but what if the user is not providing the Rest API Id directly but rather with an intrinsic function like Ref or GetAttr.

Another issue, is that multiple services can't have endpoints of the same path. Like the following:

```yml
# service A
functions:
  hello:
    handler: handler.hello
    events:
      - http: GET hello
```

# service B
functions:
  hello:
    handler: handler.hello
    events:
      - http: POST hello
```
Because the second service would attempt to create that same `hello` path/resource which is managed by service A stack. So this ends up throwing a "resource of the same parent and path already exists" error. You'd see this error too even when using this PR with one service that is trying to hook up with an existing Rest APIG that has that hello resource already defined.

A possible solution is to, again, using the SDK, only create the hello resource if it does not already exist, but then again, what if the user does not specify the Rest API Id directly?

**Comment by HyperBrain** [comment]
## Path problems / conflicts

### Theory and problem analysis

Seen from a higher point of view, each serverless service is an isolated
and integral group of functions, endpoints and resources.

When adding multiple Serverless services to one APIG API this basic concept
should be continued and followed. If we take a look at any API, the concept
of integral groups can only applied to complete hives located somewhere within
the API's path structure.

As a result (keeping consistency) Serverless services should be deployed into their
own hive in existing APIs. The hive must be seen as _owned_ by the service.

### Practical solution proposal

#### Configuration and deployment

Taking the theoretical approach into account, there have to be some additions to the current
PR implementation:

The API hive has to be configured. I would use something like
```
provider:
  name: aws
  runtime: nodejs6.10
  apiGatewayRestApiId: 6fyzt1pfpk # can be found in aws console
  apiGatewayRootPath: /myteam/v1/apis/accounts
```

This would mean that the service owns the /myteam/v1/apis/accounts path and Serverless will
prepend this path to all defined http endpoints in the service automatically. Of course it must only
do that if a common ApiId is defined. **The behavior without a shared API must not be changed!**

It is the user's responsibility to define distinct root paths for the services.

#### The root path problem

As each service sets a root path, that can have parts in common, another problem arises. These root path
parts can only be defined in one CF stack. Otherwise it would conflict due to existing resources.

Sample:
```
service A => /myteam/v1/apis/accounts
service B => /myteam/v1/apis/pages
service C => /myteam/v1/tools/editor
```

In this case the CF resources for `/myteam`, `/myteam/v1`, `/myteam/v1/tools` and `/myteam/v1/apis` must be defined in the CF template.
From a deployment perspective, the first service that deploys its part of the API has to define these. This leads to races between deployed projects.

#### The root path solution

So we need a solution for that. After some thinking, the only way to eliminate all possible races, would be to extract all CF path resources for all projects that reference the same API in their configuration 
into a separate CF stack. The stack must not be associated with a specific project, but with the API id as multiple independent projects can take part in the deployment. 

Let's stick with the example above:

If the first service is deployed, Serverless would check if an API path stack for the given API name already exists.
A suitable name for such a stack would be: `sls-api-{apiId}`.

##### Stack does not yet exist
Serverless create the API stack, and the following resources and outputs:

* RES: AWS::ApiGateway::Resource -> All path resources
* OUT (exported): Ref: AWS::ApiGateway::API

##### Stack exists

Serverless retrieves the *original* current CF template of the stack with the `getTemplate` call and
merges its resources in (deleting its own resources first). It can exactly determine its own resources
by using the apiApiGatewayRootPath.
Then it updates the API stack with the consolidated template.

##### Further common handling by Serverless

Remove all elements that are now in the API stack from the service CF template and reference the
APIG API export of the API stack in the method resources. This will also protect the API stack from deletion as long as there is any service deployed there.

### Caveats

* All path parts (up to the root) that are used by SLS services deployed to the API must be exclusively used by Serverless projects as the CF resources reside in the API stack.

* Switching a project from API integration to standard and vice versa will need a service removal before.

### Open questions

* Can the API itself be created by the participating services? This would mean that the API is still under 
full control of Serverless and nothing has to be added or created manually. How would the naming of the Api be done in that case? Normally the service name is used.

**As always: This behavior should only be triggered if a common API is set in the configuration**. Just had a crazy idea: If we had a separate API path stack (only path resources, nothing else) in general, it would solve the resource count exceeded problem automatically (just an idea - open for discussion :-) )

It would be cool if you let me know if I have some errors in my proposal or if something is in there that is not consistent or where is missed important things => is it all BS?

**Comment by HyperBrain** [comment]
I have also some thoughts about the retrieval of the root resource idea, but just forgot that to add.... I'll add these here later. Took me more time than expected to write the stuff above 😄

**Comment by HyperBrain** [comment]
@eahefnawy Please let me know what you think

**Comment by kerryhatcher** [comment]
@HyperBrain Your path logic seems sound. Generally, we would have a group of related services centered around a common namespace, and that namespace is handled by a unique team of folks and handled as a unique project. We've run into similar issues with our on-prem microservices solution. In the end, it was resolved by having a couple senior engineers that are somewhat like a domain registrar. They maintain a list of approved namespaces and keep a record of them in a wiki.

I would see something akin to that for a large serverless project (which we want to do). It's not automagic and I think the URL path conflict ("resource of the same parent and path already exists") would just a valid warning to the devs that there is a conflict. 

So if we had some _**theoretical**_ namespaces of "meta" services then:

apiGatewayRootPath: /myatc/
profile service => /myatc/profile
favortes service => /myatc/favortes

apiGatewayRootPath: /autosearch/
findcar service => /autosearch/findcar
details service => /autosearch/deatils/$vin

apiGatewayRootPath: /reviews/
dealer service => /dealer/
car service => /car/

## TL;DR; setting some common path to be prepended to all services in a project would be great.

**Comment by eahefnawy** [comment]
@HyperBrain great write up! Thanks a lot for thinking this through. I do agree with most of what you said, and while I think this could would solve the problems listed above, I'm worried more problems would come out of this not-so-simple solution. For example, I'm worried the deployment speed would suffer significantly by introducing a new stack to be created/updated. I think our current deployment speed is already not so great. Do you think there's a way to get it working without introducing a new stack?

**Comment by HyperBrain** [comment]
I thought quite a long time about solving the path problem. This was the only solution I came to that would solve it without the risk of running into problems. The major problem is imo, that the services are distinct and do not know which resources are used in common, and which aren't. For CF a conflict is a terminating condition, that will abort the deployment if a resource already exists (at least as far as I know). I fully agree that deployment performance is a major point and should not be affected by features that are not used by the majority of users.

If the api root (first path element) - as shown in @kerryhatcher 's comment above - has to be different per service (declared as fixed requirement for the functionality), it would work without a separate stack, but in that case *it's the user's responsibility to prevent clashes*. Such a definition must be well documented and cannot be checked for compliance automatically by Serverless. That way it could go into the core framework.

For the full fledged solution, a way to go would be to put the common API support into a separate plugin (maybe owned by Serverless to make sure that it always works with the core framework). For these users who make use of it, the decreased deployment speed would be ok in favor of the feature - all others would not experience any worsening.

I think, in the end it comes to the decision - have it with the fixed root path requirement in core, or have it fully configurable in a separate plugin.

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -146,22 +146,11 @@ class AwsProvider {
         f()
           .then(resolve)
           .catch((e) => {
-            const err = e;
-            if (err.statusCode === 429) {
+            if (e.statusCode === 429) {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.js]: Can we quick-fix this bug in a separate PR since it's unrelated to the shared API Gateway PR (I can do that since I broke it recently 😬).

**Comment by potocnika** [comment]
What's needed to fix the failing builds? Really looking forward to this feature being merged. Can anyone help?

**Comment by haines** [comment]
Re: the path problem, what if it were possible to specify (and require) that some of the API gateway resources will already exist, as well as the gateway?

Our use case is that we would like to manage the API gateway in Terraform and deploy into it with Serverless (this allows us, among other things, to hook in a custom domain name). In that case, it would be simple for us to manage the "root path" resources along with the gateway.

I wonder if it would be sufficient to make that a documented caveat of this feature (if multiple services deployed to the same gateway are to share a root path, then those common resources need to be managed externally)?

**Comment by ynnr85** [comment]
Any update about it?

**Comment by hgiasac** [comment]
In my real-world project, because of bad API design, we encounter much conflict of predefined resources

```
# Service A:
/users/{userId}
/users/{userId}/posts
```
```
# Service B:
/users/jobs
/users/{userId}/jobs
```
So we need more than 1 root resource. The applicable solution is, define many deployed resource as you can:

```
provider:
  apiGateway:
    restApiId: abcdef
    restApiRootResourceId: abcxyz
    restApiResources:
      - name: Users
        path: /users
        resourceId: { Ref: ApiResourceUsers }
      
      - name: UserDetail
        path: /users/{userId}
        resourceId: { Ref: ApiResourceUserDetail }

functions:
  userPosts:
    handler: handler.userPosts
    events:
      - http: GET /users/{userId}/posts

  userInfo:
    handler: handler.userInfo
    events:
      - http: GET /users/info

  tags:
    handler: handler.tags
    events:
      - http: GET /tags
```

Under the hood, the compiler auto generate resource config from root path, then children path reference the parentId. In this case, with predefined resources, we can map them with parent paths. If there are other paths that need to be auto-generated, the compiler can still do it. (In this case, path /tags are auto-created, in the same serverless.yml config file )

I can't wait for this PR to be merged, so I checkout @eahefnawy code with some modification. And it run smoothly.

**Comment by eahefnawy** [comment]
thanks for sharing your thoughts @hgiasac 😊 ... curious what modification you've added to this PR. We'd probably need them here too ✨

**Comment by hgiasac** [comment]
@eahefnawy I create a branch in my forked repo, feel free to check it out 
https://github.com/hgiasac/serverless/tree/sls-shared-api

The config is same as my early post. I change it from the requested change of @pmuens 
You can add many root resource paths. If there is conflict path, we will know specific path where resourceId need to be assigned

**Comment by pmuens** [comment]
> @eahefnawy I create a branch in my forked repo, feel free to check it out
https://github.com/hgiasac/serverless/tree/sls-shared-api

> The config is same as my early post. I change it from the requested change of @pmuens
You can add many root resource paths. If there is conflict path, we will know specific path where resourceId need to be assigned

Great! Thanks for PRing the changes @hgiasac 👍

**Comment by pkubat** [comment]
These PR changes do pass test locally with the current master.   What is the best means to get this PR updated with the current master? (NOTE: The lint fails due to needing an extra newline in providers). So this can be merged? Thanks!

**Comment by j0k3r** [comment]
@pkubat you can create a new branch from `shared-api` branch, fix the problem and then open a new PR targetting the `shared-api` branch. When it'll be merged, test will then be ok here.

**Comment by Christoph-Schabert** [comment]
Would it not be simple to write a custom resource for the API gateway resource resource?
This would check if the path is already created and remove the path if nothing is attached?
This would reduce the complexity a lot and make it simpler until cloudformatation support already existing resources.

I try to create this plugin but I am not very familiar with node is

**Comment by hgiasac** [comment]
So have someone checked out my [branch](https://github.com/serverless/serverless/pull/3934#issuecomment-326749622)? Do you have any feedback about my [idea](https://github.com/serverless/serverless/pull/3934#issuecomment-326396191) ? Can I create another PR too. Thanks

**Comment by abhilash80** [comment]
Getting the following error while deploying with a custom domain using the serverless-domain-manager plugin.

`Template format error: Unresolved resource dependencies [ApiGatewayRestApi] in the Resources block of the template` 

This is the stack trace:

``` 
Error: The CloudFormation template is invalid: Template format error: Unresolved resource dependencies [ApiGatewayRestApi] in the Resources block of the template
    at provider.request.catch (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/plugins/aws/deploy/lib/validateTemplate.js:25:13)
From previous event:
    at AwsDeploy.validateTemplate (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/plugins/aws/deploy/lib/validateTemplate.js:20:12)
From previous event:
    at AwsDeploy.BbPromise.bind.then (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/plugins/aws/deploy/index.js:117:39)
From previous event:
    at Object.aws:deploy:deploy:validateTemplate [as hook] (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/plugins/aws/deploy/index.js:113:10)
    at BbPromise.reduce (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/classes/PluginManager.js:242:55)
From previous event:
    at PluginManager.invoke (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/classes/PluginManager.js:242:22)
    at PluginManager.spawn (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/classes/PluginManager.js:254:17)
    at AwsDeploy.BbPromise.bind.then (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/plugins/aws/deploy/index.js:91:48)
From previous event:
    at Object.deploy:deploy [as hook] (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/plugins/aws/deploy/index.js:87:10)
    at BbPromise.reduce (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/classes/PluginManager.js:242:55)
From previous event:
    at PluginManager.invoke (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/classes/PluginManager.js:242:22)
    at PluginManager.run (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/classes/PluginManager.js:261:17)
    at variables.populateService.then (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/Serverless.js:99:33)
    at runCallback (timers.js:781:20)
    at tryOnImmediate (timers.js:743:5)
    at processImmediate [as _immediateCallback] (timers.js:714:5)
From previous event:
    at Serverless.run (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/lib/Serverless.js:86:74)
    at serverless.init.then (/Users/ap/Documents/Code/serverlessfw-experiments/serverless-sls-shared-api/bin/serverless:39:50)
    at <anonymous>
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.6.0
     Serverless Version:     1.23.0


```

**This is what my serverless.yml file looks like:**

```
service: serverlessfw-dtsvc

plugins:
  - authorizer-arn-from-id-plugin
  - serverless-domain-manager

custom:
  appconfig: ${file(../common/appconfig.yml)}
  customDomain:
    domainName: api.######.com
    basePath: ''
    stage: ${self:provider.stage}
    createRoute53Record: true  

provider:
  name: aws
  runtime: ${self:custom.appconfig.LambdaRuntime}
  memorySize: ${self:custom.appconfig.LambdaMemorySize}
  deploymentBucket: ${self:custom.appconfig.DeployS3Bucket}
  versionFunctions: ${self:custom.appconfig.Lambda_Versioning}
  role: 
    Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-lambda-iam-role
  apiGateway:
    restApiId: 
      Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-apigw-id
    restApiRootResourceId:
      Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-apigw-root-resource-id
  stackTags: ${self:custom.appconfig.StandardTags}

functions:
  dt-svc:
    handler: handler.getServerDateTime
    events:
      - http: 
          path: serverdate
          method: get
          authorizer: 
            type: CUSTOM
            authorizerId: 
              Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-apigw-authorizer-id
    vpc:
      securityGroupIds: 
        - Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-lambda-sg
      subnetIds: ${self:custom.appconfig.VPCInternalSubnetIDs}
```

Not sure if i have set this up incorrectly?

**Comment by HyperBrain** [comment]
@abhilash80 I think this is your problem:
```
 apiGateway:
    restApiId: 
      Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-apigw-id
    restApiRootResourceId:
      Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-apigw-root-resource-id
```
That means that the stack does not contain an `ApiGatewayRestApi` resource nor an `ApiGatewayRootResource` on which the plugin might depend. I'm pretty sure that this has to be fixed in the plugin, as the behavior to set a custom APIG API is quite new. BTW: the aws-alias plugin does have the same problem.

So I'd expect that it will take some time until all plugins that modify the APIG parts in the template are all adapted properly. Until then the functionality is only guaranteed to run with vanilla Serverless.

@eahefnawy @pkubat @hgiasac Any opinion here?

**Comment by hgiasac** [comment]
@abhilash80 you can check with hard-coded API Gateway ID first. We don't know that referenced AGW resource have exists in CloudFormation stack yet. Did you create `AWS::ApiGateway::RestApi` API resource and output it before?

**Comment by abhilash80** [comment]
@HyperBrain @hgiasac 

Appreciate the quick response!

I get the same error with the hard coded values as well - following is the yaml snippet:

```
  name: aws
  runtime: ${self:custom.appconfig.LambdaRuntime}
  memorySize: ${self:custom.appconfig.LambdaMemorySize}
  deploymentBucket: ${self:custom.appconfig.DeployS3Bucket}
  versionFunctions: ${self:custom.appconfig.Lambda_Versioning}
  role: 
    Fn::ImportValue: ${self:custom.appconfig.CommonStackName}-lambda-iam-role
  apiGateway:
    restApiId: ndxyt6kqv6
    restApiRootResourceId: 66ob90flgk
  stackTags: ${self:custom.appconfig.StandardTags}
```

The API Gateway has been created in a separate stack and the id's exported as output:

```
  ApiGatewayRestApi:
    Type: AWS::ApiGateway::RestApi
    Properties:
      Name: serverlessfw-apigw

Outputs:
  ServerlessFWAPIGWID:
    Description: API Gateway API ID
    Value: 
      Ref: ApiGatewayRestApi
    Export:
      Name:
        Fn::Sub: "${AWS::StackName}-apigw-id"

  # API Gateway Root Resource ID
  ServerlessFWAPIGWRootResourceID:
    Description: API GW Root Resource ID
    Value: !GetAtt ApiGatewayRestApi.RootResourceId
    Export:
      Name:
        Fn::Sub: "${AWS::StackName}-apigw-root-resource-id"
```

The stack is already provisioned and I can see the exported values in the console:

image

**Comment by HyperBrain** [comment]
@abhilash80 I strongly assume that the domain name plugin depends on an existing APIGateway resource IN the serverless stack (as it needs to reference it) - it injects the domain name mappings into the template and thus creates invalid references (as each other plugin will do that is not yet adapted).

Can you try if it works when you remove the plugin? If this solves the problem, this task has to be explicitly flagged as BREAKING, @horike37 and must be targeted for a major release.

**Comment by hgiasac** [comment]
I think the issue is in custom domain plugin, because it reference default API Gateway resource that is created automatically by serverless framework. Meanwhile we use another AGW Resource name, default ApiGatewayRestApi resource does not exists in current stack

**Comment by abhilash80** [comment]
I tried it out and it looks like the problem occurs when the cross-domain plugin is used. Do you guys have any suggestions or workarounds?

I will try starting a thread with the plugin to see what they have to say.

**Comment by HyperBrain** [comment]
@abhilash80 I think ALL plugins that modify APIG resources will not work anymore. Imo it will take some time to get them all updated (because that is most likely not that easy - especially as e.g. the owner of the domain mapping resource is then unclear). What if a different service that uses the same API also adds a domain name mapping? This might collide.

**Comment by hgiasac** [comment]
@abhilash80 You must use the plugin in master serverless.yml project that define `ApiGatewayRestApi` resource. No need to use it in children projects

**Comment by HyperBrain** [comment]
@hgiasac But that only works for plugins that are stateless, i.e. do only static things. The majority of plugins will do their operations depending on the project/service they are run in.
However we should streamline a discussion about these incompatibilities and proper solution ideas in this thread - together with the plugin authors. It does not make sense to come up with a solution that fits for plugin A but does not work with the other plugins.
I know that this is a tough thing, but imo it is the only way to resolve it _before_ this gets into master (honestly, I even think that this is kind of blocker for the merge as people will start using the feature as soon as it's in - with the plugins they are currently using).

**Comment by hgiasac** [comment]
@HyperBrain you're right. Plugins need to be update to work if this PR is merged
@abhilash80 Have you tried with my suggestion? Does it work? Just remove the plugin from child AGW projects

**Comment by abhilash80** [comment]
@hgiasac Thanks for the suggestion!

I get the following error when I try to move the custom domain plugin into the master serverless.yml file:

`Error: Cannot find AWS::ApiGateway::Deployment Try running sls create_domain first.`

And I had run the create-domain command before so the domain is provisioned and initialized.

**Comment by abhilash80** [comment]
@HyperBrain Since you were involved with the [Alias Plugin](https://github.com/HyperBrain/serverless-aws-alias), wanted to confirm if it's compliant with the code in this PR?

Thanks

**Comment by aalarcong** [comment]
Hello.
Whats the status of this feature?

**Comment by a-suenami** [comment]
It's been on also my mind. I want this feature.

**Comment by fazelmk** [comment]
Does it work on this branch? 
Tried to use with apiId and didn't work. If it works how I use it?

**Comment by hgiasac** [comment]
@fazelmk You should change to this branch #4247. This feature will be official in version 1.26

**Comment by horike37** [comment]
Closing this issue for now since this work in progress on #4247.
